# TABLE OF CONTENT:

# 0. Upload data set and libraries

# 1. One hot encoding

## 1.1 Primary Types

## 1.2 Location

# 2. Add the arrest rate/district in encoding table

# 3. Machine Learning Algorithms 

## 3.1 Logistic Regression

## 3.2 Random Forest 

## 3.3 Decision Tree

# 4. Test the RF algorithm with new data 




# 0. Upload dataset and libraries

In [1]:
import pandas as pd
import numpy as np
import plotly.plotly as py
import plotly.graph_objs as go
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.plotly as py
import cufflinks as cf
cf.go_offline()
from ipywidgets import interact

from scipy import stats
import scipy.stats as stats
import statsmodels as sm

In [2]:
import warnings
warnings.simplefilter('default')

In [4]:
pd.set_option('display.max_columns',100)

In [5]:
chicago_cmr = pd.read_csv('/Users/adrientalbot/Documents/final-project/data/chicago_clean_month.csv')

In [6]:
chicago_formatted = chicago_cmr.drop(columns = ['Latitude','Longitude'])

In [7]:
chicago_formatted.head()

Unnamed: 0                 Date  IUCR                Primary Type  \
0           0  2001-01-01 00:00:00  0840                       THEFT   
1           1  2001-01-01 00:00:00  1752  OFFENSE INVOLVING CHILDREN   
2           2  2001-01-01 00:00:00  0810                       THEFT   
3           3  2001-01-01 00:00:00  0840                       THEFT   
4           4  2001-01-01 00:00:00  1754  OFFENSE INVOLVING CHILDREN   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE       0         0   
1   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE       0         0   
2                       OVER $500            APARTMENT       0         0   
3   FINANCIAL ID THEFT: OVER $300            RESIDENCE       0         0   
4  AGG SEX ASSLT OF CHILD FAM MBR            RESIDENCE       0         0   

   Beat  District  Year                       Location YearMonth  
0   631       6.0  2001  (41.748623027, -87.608654109)   2001-01  
1   624       6.0  2001  (41.754802305, -87.598146873)   2001-01  
2  2332      19.0  2001  (41.938952432, -87.638163582)   2001-01  
3   624       6.0  2001  (41.752061911, -87.607827503)   2001-01  
4  1013      10.0  2001  (41.850385805, -87.715107802)   2001-01

# 1. One hot encoding 

## 1.1 One hot encoding for Primary Types

In [8]:
primary_type_count = chicago_formatted['Primary Type'].value_counts()
df_primary_type_count = pd.DataFrame(primary_type_count).reset_index().rename({"index": "primary type", "Primary Type": "count"}, axis = 1)
top_23_primary = df_primary_type_count[df_primary_type_count['count'] > 5000]
top_23_primary_lst = list(top_23_primary['primary type'].unique())

In [9]:
primary_type_count

THEFT                                1425163
BATTERY                              1244942
CRIMINAL DAMAGE                       777380
NARCOTICS                             707206
ASSAULT                               424171
OTHER OFFENSE                         422279
BURGLARY                              390379
MOTOR VEHICLE THEFT                   314879
DECEPTIVE PRACTICE                    257647
ROBBERY                               257343
CRIMINAL TRESPASS                     195041
WEAPONS VIOLATION                      72432
PROSTITUTION                           67919
PUBLIC PEACE VIOLATION                 48135
OFFENSE INVOLVING CHILDREN             43994
CRIM SEXUAL ASSAULT                    26408
SEX OFFENSE                            24384
INTERFERENCE WITH PUBLIC OFFICER       15574
GAMBLING                               14340
LIQUOR LAW VIOLATION                   14002
ARSON                                  11210
HOMICIDE                                9608
KIDNAPPING

In [10]:
lst_primary_types = ['THEFT', 'OFFENSE INVOLVING CHILDREN', 'ASSAULT',
       'CRIM SEXUAL ASSAULT', 'SEX OFFENSE', 'DECEPTIVE PRACTICE',
       'CRIMINAL DAMAGE', 'BATTERY',
       'MOTOR VEHICLE THEFT', 'BURGLARY', 'WEAPONS VIOLATION',
       'CRIMINAL TRESPASS', 'ROBBERY', 'NARCOTICS',
       'LIQUOR LAW VIOLATION', 'PUBLIC PEACE VIOLATION',
       'INTERFERENCE WITH PUBLIC OFFICER', 'HOMICIDE', 'GAMBLING',
       'KIDNAPPING', 'PROSTITUTION', 'ARSON']

In [11]:
chicago_formatted_primary = chicago_formatted[chicago_formatted['Primary Type'].isin(lst_primary_types)]

In [12]:
for primary_type in lst_primary_types:
    chicago_formatted_primary[primary_type] = 0

In [13]:
for primary_type in lst_primary_types:
    chicago_formatted_primary[primary_type] = np.where(chicago_formatted_primary['Primary Type'].str.contains(primary_type), 1, 0)


In [14]:
chicago_formatted_primary.head()

Unnamed: 0                 Date  IUCR                Primary Type  \
0           0  2001-01-01 00:00:00  0840                       THEFT   
1           1  2001-01-01 00:00:00  1752  OFFENSE INVOLVING CHILDREN   
2           2  2001-01-01 00:00:00  0810                       THEFT   
3           3  2001-01-01 00:00:00  0840                       THEFT   
4           4  2001-01-01 00:00:00  1754  OFFENSE INVOLVING CHILDREN   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE       0         0   
1   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE       0         0   
2                       OVER $500            APARTMENT       0         0   
3   FINANCIAL ID THEFT: OVER $300            RESIDENCE       0         0   
4  AGG SEX ASSLT OF CHILD FAM MBR            RESIDENCE       0         0   

   Beat  District  Year                       Location YearMonth  THEFT  \
0   631       6.0  2001  (41.748623027, -87.608654109)   2001-01      1   
1   624       6.0  2001  (41.754802305, -87.598146873)   2001-01      0   
2  2332      19.0  2001  (41.938952432, -87.638163582)   2001-01      1   
3   624       6.0  2001  (41.752061911, -87.607827503)   2001-01      1   
4  1013      10.0  2001  (41.850385805, -87.715107802)   2001-01      0   

   OFFENSE INVOLVING CHILDREN  ASSAULT  CRIM SEXUAL ASSAULT  SEX OFFENSE  \
0                           0        0                    0            0   
1                           1        0                    0            0   
2                           0        0                    0            0   
3                           0        0                    0            0   
4                           1        0                    0            0   

   DECEPTIVE PRACTICE  CRIMINAL DAMAGE  BATTERY  MOTOR VEHICLE THEFT  \
0                   0                0        0                    0   
1                   0                0        0                    0   
2                   0                0        0                    0   
3                   0                0        0                    0   
4                   0                0        0                    0   

   BURGLARY  WEAPONS VIOLATION  CRIMINAL TRESPASS  ROBBERY  NARCOTICS  \
0         0                  0                  0        0          0   
1         0                  0                  0        0          0   
2         0                  0                  0        0          0   
3         0                  0                  0        0          0   
4         0                  0                  0        0          0   

   LIQUOR LAW VIOLATION  PUBLIC PEACE VIOLATION  \
0                     0                       0   
1                     0                       0   
2                     0                       0   
3                     0                       0   
4                     0                       0   

   INTERFERENCE WITH PUBLIC OFFICER  HOMICIDE  GAMBLING  KIDNAPPING  \
0                                 0         0         0           0   
1                                 0         0         0           0   
2                                 0         0         0           0   
3                                 0         0         0           0   
4                                 0         0         0           0   

   PROSTITUTION  ARSON  
0             0      0  
1             0      0  
2             0      0  
3             0      0  
4             0      0

In [15]:
chicago_formatted_primary['Primary Type'].unique()

array(['THEFT', 'OFFENSE INVOLVING CHILDREN', 'ASSAULT',
       'CRIM SEXUAL ASSAULT', 'SEX OFFENSE', 'DECEPTIVE PRACTICE',
       'CRIMINAL DAMAGE', 'BATTERY', 'MOTOR VEHICLE THEFT', 'BURGLARY',
       'WEAPONS VIOLATION', 'CRIMINAL TRESPASS', 'ROBBERY', 'NARCOTICS',
       'LIQUOR LAW VIOLATION', 'PUBLIC PEACE VIOLATION',
       'INTERFERENCE WITH PUBLIC OFFICER', 'HOMICIDE', 'GAMBLING',
       'KIDNAPPING', 'PROSTITUTION', 'ARSON'], dtype=object)

## 1.2 One encoding for location description

In [16]:
location_count = chicago_formatted_primary['Location Description'].value_counts()

In [17]:
df_location_count = pd.DataFrame(location_count).reset_index().rename({"index": "location description", "Location Description": "count"}, axis = 1)
top_23_location = df_location_count[df_location_count['count'] > 35000]
top_23_location_lst = list(top_23_location['location description'].unique())

In [18]:
lst_location = ['STREET',
 'RESIDENCE',
 'SIDEWALK',
 'APARTMENT',
 'PARKING LOT/GARAGE(NON.RESID.)',
 'ALLEY',
 'SCHOOL, PUBLIC, BUILDING',
 'RESIDENCE-GARAGE',
 'SMALL RETAIL STORE',
 'RESIDENCE PORCH/HALLWAY',
 'RESTAURANT',
 'VEHICLE NON-COMMERCIAL',
 'GROCERY FOOD STORE',
 'DEPARTMENT STORE',
 'GAS STATION',
 'RESIDENTIAL YARD (FRONT/BACK)',
 'CHA PARKING LOT/GROUNDS',
 'PARK PROPERTY',
 'COMMERCIAL / BUSINESS OFFICE',
 'CTA PLATFORM',
 'BAR OR TAVERN']

In [19]:
chicago_formatted_primary_location = chicago_formatted_primary[chicago_formatted_primary['Location Description'].isin(lst_location)]

In [20]:
chicago_formatted_primary_location['Primary Type'].unique()

array(['THEFT', 'OFFENSE INVOLVING CHILDREN', 'ASSAULT',
       'CRIM SEXUAL ASSAULT', 'SEX OFFENSE', 'DECEPTIVE PRACTICE',
       'CRIMINAL DAMAGE', 'BATTERY', 'MOTOR VEHICLE THEFT', 'BURGLARY',
       'WEAPONS VIOLATION', 'CRIMINAL TRESPASS', 'ROBBERY', 'NARCOTICS',
       'LIQUOR LAW VIOLATION', 'PUBLIC PEACE VIOLATION', 'GAMBLING',
       'PROSTITUTION', 'KIDNAPPING', 'INTERFERENCE WITH PUBLIC OFFICER',
       'ARSON', 'HOMICIDE'], dtype=object)

In [21]:
for location in lst_location:
    chicago_formatted_primary_location[location] = 0

In [22]:
for location in lst_location:
    chicago_formatted_primary_location[location] = np.where(chicago_formatted_primary_location['Location Description'].str.contains(location), 1, 0)


/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: UserWarning:

This pattern has match groups. To actually get the groups, use str.extract.



In [23]:
chicago_formatted_primary_location.head()

Unnamed: 0                 Date  IUCR                Primary Type  \
0           0  2001-01-01 00:00:00  0840                       THEFT   
1           1  2001-01-01 00:00:00  1752  OFFENSE INVOLVING CHILDREN   
2           2  2001-01-01 00:00:00  0810                       THEFT   
3           3  2001-01-01 00:00:00  0840                       THEFT   
4           4  2001-01-01 00:00:00  1754  OFFENSE INVOLVING CHILDREN   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE       0         0   
1   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE       0         0   
2                       OVER $500            APARTMENT       0         0   
3   FINANCIAL ID THEFT: OVER $300            RESIDENCE       0         0   
4  AGG SEX ASSLT OF CHILD FAM MBR            RESIDENCE       0         0   

   Beat  District  Year                       Location YearMonth  THEFT  \
0   631       6.0  2001  (41.748623027, -87.608654109)   2001-01      1   
1   624       6.0  2001  (41.754802305, -87.598146873)   2001-01      0   
2  2332      19.0  2001  (41.938952432, -87.638163582)   2001-01      1   
3   624       6.0  2001  (41.752061911, -87.607827503)   2001-01      1   
4  1013      10.0  2001  (41.850385805, -87.715107802)   2001-01      0   

   OFFENSE INVOLVING CHILDREN  ASSAULT  CRIM SEXUAL ASSAULT  SEX OFFENSE  \
0                           0        0                    0            0   
1                           1        0                    0            0   
2                           0        0                    0            0   
3                           0        0                    0            0   
4                           1        0                    0            0   

   DECEPTIVE PRACTICE  CRIMINAL DAMAGE  BATTERY  MOTOR VEHICLE THEFT  \
0                   0                0        0                    0   
1                   0                0        0                    0   
2                   0                0        0                    0   
3                   0                0        0                    0   
4                   0                0        0                    0   

   BURGLARY  WEAPONS VIOLATION  CRIMINAL TRESPASS  ROBBERY  NARCOTICS  \
0         0                  0                  0        0          0   
1         0                  0                  0        0          0   
2         0                  0                  0        0          0   
3         0                  0                  0        0          0   
4         0                  0                  0        0          0   

   LIQUOR LAW VIOLATION  PUBLIC PEACE VIOLATION  \
0                     0                       0   
1                     0                       0   
2                     0                       0   
3                     0                       0   
4                     0                       0   

   INTERFERENCE WITH PUBLIC OFFICER  HOMICIDE  GAMBLING  KIDNAPPING  \
0                                 0         0         0           0   
1                                 0         0         0           0   
2                                 0         0         0           0   
3                                 0         0         0           0   
4                                 0         0         0           0   

   PROSTITUTION  ARSON  STREET  RESIDENCE  SIDEWALK  APARTMENT  \
0             0      0       0          1         0          0   
1             0      0       0          1         0          0   
2             0      0       0          0         0          1   
3             0      0       0          1         0          0   
4             0      0       0          1         0          0   

   PARKING LOT/GARAGE(NON.RESID.)  ALLEY  SCHOOL, PUBLIC, BUILDING  \
0                               0      0                         0   
1                               0      0                      

In [24]:
one_hot_table = chicago_formatted_primary_location.drop(['Description','Location Description', 'Beat', 'Year','Unnamed: 0','IUCR'], axis = 1)

# 2. Add arrest rate/district

In [25]:
chicago_crime_counts = pd.DataFrame(one_hot_table['District'].value_counts()).reset_index()

In [26]:
chicago_crime_counts.head()

index  District
0    8.0    382462
1   11.0    370906
2    7.0    343966
3   25.0    330273
4    6.0    328372

In [27]:
chicago_crime_counts = chicago_crime_counts.rename(columns={'index': 'District', 'District': 'Crime_Counts'})
chicago_crime_counts.head()

District  Crime_Counts
0       8.0        382462
1      11.0        370906
2       7.0        343966
3      25.0        330273
4       6.0        328372

In [28]:
chicago_arrest_counts = pd.DataFrame(one_hot_table[one_hot_table['Arrest'] == 1]['District'].value_counts()).reset_index()

In [29]:
chicago_arrest_counts = chicago_arrest_counts.rename(columns={'index': 'District', 'District': 'Crime_Counts'})
chicago_arrest_counts.head()

District  Crime_Counts
0      11.0        167204
1      15.0        111710
2       7.0        101143
3      25.0         97661
4       8.0         95528

In [30]:
chicago_map = chicago_crime_counts.merge(chicago_arrest_counts, on='District', how='left')
chicago_map = chicago_map.rename({'Crime_Counts_x': 'total crimes', 'Crime_Counts_y': 'total arrests'}, axis = 1)

In [31]:
chicago_map['arrest rate'] = chicago_map['total arrests'] / chicago_map['total crimes']

In [32]:
chicago_map.head()

District  total crimes  total arrests  arrest rate
0       8.0        382462          95528     0.249771
1      11.0        370906         167204     0.450799
2       7.0        343966         101143     0.294049
3      25.0        330273          97661     0.295698
4       6.0        328372          91223     0.277804

In [33]:
ml_df = one_hot_table.merge(chicago_map, on = 'District', how ='left')

In [34]:
ml_df = ml_df.drop(['Primary Type', 'Date' , 'Location', 'YearMonth', 'total crimes' ,'total arrests', 'District'], axis = 1)

In [35]:
ml_df.head()

Arrest  Domestic  THEFT  OFFENSE INVOLVING CHILDREN  ASSAULT  \
0       0         0      1                           0        0   
1       0         0      0                           1        0   
2       0         0      1                           0        0   
3       0         0      1                           0        0   
4       0         0      0                           1        0   

   CRIM SEXUAL ASSAULT  SEX OFFENSE  DECEPTIVE PRACTICE  CRIMINAL DAMAGE  \
0                    0            0                   0                0   
1                    0            0                   0                0   
2                    0            0                   0                0   
3                    0            0                   0                0   
4                    0            0                   0                0   

   BATTERY  MOTOR VEHICLE THEFT  BURGLARY  WEAPONS VIOLATION  \
0        0                    0         0                  0   
1        0                    0         0                  0   
2        0                    0         0                  0   
3        0                    0         0                  0   
4        0                    0         0                  0   

   CRIMINAL TRESPASS  ROBBERY  NARCOTICS  LIQUOR LAW VIOLATION  \
0                  0        0          0                     0   
1                  0        0          0                     0   
2                  0        0          0                     0   
3                  0        0          0                     0   
4                  0        0          0                     0   

   PUBLIC PEACE VIOLATION  INTERFERENCE WITH PUBLIC OFFICER  HOMICIDE  \
0                       0                                 0         0   
1                       0                                 0         0   
2                       0                                 0         0   
3                       0                                 0         0   
4                       0                                 0         0   

   GAMBLING  KIDNAPPING  PROSTITUTION  ARSON  STREET  RESIDENCE  SIDEWALK  \
0         0           0             0      0       0          1         0   
1         0           0             0      0       0          1         0   
2         0           0             0      0       0          0         0   
3         0           0             0      0       0          1         0   
4         0           0             0      0       0          1         0   

   APARTMENT  PARKING LOT/GARAGE(NON.RESID.)  ALLEY  SCHOOL, PUBLIC, BUILDING  \
0          0                               0      0                         0   
1          0                               0      0                         0   
2          1                               0      0                         0   
3          0                               0      0                         0   
4          0                               0      0                         0   

   RESIDENCE-GARAGE  SMALL RETAIL STORE  RESIDENCE PORCH/HALLWAY  RESTAURANT  \
0                 0                   0                        0           0   
1                 0                   0                        0           0   
2                 0                   0                        0           0   
3                 0                   0                        0           0   
4                 0                   0                        0           0   

   VEHICLE NON-COMMERCIAL  GROCERY FOOD STORE  DEPARTMENT STORE  GAS STATION  \
0                       0                   0                 0            0   
1                       0                   0                 0            0   
2                       0                   0                 0            0   
3                       0                   0                 0            0   
4                       0                   0                 0            0   

   RESIDEN

In [36]:
#ml_df.to_csv(r'/Users/adrientalbot/Documents/final-project/data/dataset_ml.csv')

In [37]:
ml_df.corr()

Arrest  Domestic     THEFT  \
Arrest                            1.000000 -0.066519 -0.229230   
Domestic                         -0.066519  1.000000 -0.194361   
THEFT                            -0.229230 -0.194361  1.000000   
OFFENSE INVOLVING CHILDREN       -0.012549  0.098732 -0.051168   
ASSAULT                          -0.036788  0.081182 -0.166578   
CRIM SEXUAL ASSAULT              -0.017362  0.002101 -0.038055   
SEX OFFENSE                       0.003457 -0.008293 -0.036162   
DECEPTIVE PRACTICE               -0.043795 -0.064002 -0.108983   
CRIMINAL DAMAGE                  -0.183550 -0.055353 -0.227458   
BATTERY                          -0.066975  0.474042 -0.304414   
MOTOR VEHICLE THEFT              -0.103372 -0.090412  0.396979   
BURGLARY                         -0.133259 -0.095854 -0.155820   
WEAPONS VIOLATION                 0.125423 -0.042331 -0.066663   
CRIMINAL TRESPASS                 0.142116 -0.041279 -0.095442   
ROBBERY                          -0.090477 -0.071596 -0.125923   
NARCOTICS                         0.577193 -0.143588 -0.220804   
LIQUOR LAW VIOLATION              0.067570 -0.016829 -0.025958   
PUBLIC PEACE VIOLATION            0.076677 -0.021375 -0.052598   
INTERFERENCE WITH PUBLIC OFFICER  0.072200 -0.019159 -0.030477   
HOMICIDE                          0.012488 -0.007253 -0.020307   
GAMBLING                          0.076120 -0.018986 -0.029165   
KIDNAPPING                       -0.012865  0.013380 -0.019902   
PROSTITUTION                      0.172202 -0.042822 -0.065713   
ARSON                            -0.014889 -0.009548 -0.025463   
STREET                           -0.009873 -0.149712  0.166195   
RESIDENCE                        -0.144873  0.171091 -0.125655   
SIDEWALK                          0.181801 -0.053358 -0.161130   
APARTMENT                        -0.082492  0.267331 -0.128707   
PARKING LOT/GARAGE(NON.RESID.)         NaN       NaN       NaN   
ALLEY                             0.063390 -0.032963 -0.051813   
SCHOOL, PUBLIC, BUILDING          0.008114 -0.057742 -0.030501   
RESIDENCE-GARAGE                 -0.079096 -0.054656 -0.045305   
SMALL RETAIL STORE               -0.002635 -0.054290  0.105821   
RESIDENCE PORCH/HALLWAY           0.007311  0.001635 -0.021363   
RESTAURANT                       -0.022129 -0.046691  0.038398   
VEHICLE NON-COMMERCIAL            0.002478 -0.012939  0.034636   
GROCERY FOOD STORE                0.075140 -0.046668  0.111534   
DEPARTMENT STORE                  0.081141 -0.047493  0.160368   
GAS STATION                       0.024258 -0.035938  0.009770   
RESIDENTIAL YARD (FRONT/BACK)          NaN       NaN       NaN   
CHA PARKING LOT/GROUNDS           0.092537 -0.023834 -0.050421   
PARK PROPERTY                     0.010779 -0.027041 -0.002138   
COMMERCIAL / BUSINESS OFFICE     -0.035717 -0.030290  0.029767   
CTA PLATFORM                      0.056557 -0.028185 -0.019498   
BAR OR TAVERN                    -0.011706 -0.025439  0.025520   
arrest rate                       0.143481  0.047119 -0.119112   

                                  OFFENSE INVOLVING CHILDREN   ASSAULT  \
Arrest                                             -0.012549 -0.036788   
Domestic                                            0.098732  0.081182   
THEFT                                              -0.051168 -0.166578   
OFFENSE INVOLVING CHILDREN                          1.000000 -0.023611   
ASSAULT                                            -0.023611  1.000000   
CRIM SEXUAL ASSAULT                                -0.005394  0.228449   
SEX OFFENSE                                        -0.005126 -0.016687   
DECEPTIVE PRACTICE                                 -0.015448 -0.050290   
CRIMINAL DAMAGE                                    -0.032241 -0.104960   
BATTERY                                            -0.043149 -0.140471   
MOTOR VEHICLE THEFT                                -0.020313 -0.066128   
BURGLARY                                     

# 3. Which machine learning algorithm to use?

## 3.1 Logistic Regression

### No undersampling

In [38]:
from sklearn.model_selection import train_test_split
y = ml_df['Arrest']
x = ml_df.drop(['Arrest'], axis = 1)
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42, stratify=y)

In [39]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train,y_train)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.



LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [40]:
model.score(X_test,y_test)

0.8711142038625251

In [41]:
y_pred = model.predict(X_test)

In [42]:
from sklearn.metrics import confusion_matrix, classification_report

In [43]:
confusion_matrix(y_test, y_pred)

array([[763436,  27313],
       [115303, 200478]])

### cross validation

In [44]:
from sklearn.model_selection import cross_val_score

In [45]:
cv = cross_val_score(LogisticRegression(),X_test,y_test,cv=10)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logis

In [46]:
np.mean(cv)

0.8707292177583466

### with undersampling

In [47]:
ml_df['Arrest'].value_counts()

0    3953747
1    1578903
Name: Arrest, dtype: int64

In [48]:
arrest_false = ml_df[ml_df['Arrest'] == 0]
arrest_true = ml_df[ml_df['Arrest'] == 1]

In [49]:
count_arrest_false, count_arrest_true = ml_df['Arrest'].value_counts()

In [50]:
count_arrest_false

3953747

In [51]:
count_arrest_true

1578903

In [52]:
# Divide data from class 1 in order to have same number of values as class 0
arrest_false_under = arrest_false.sample(count_arrest_true)
df_test_under = pd.concat([arrest_false_under, arrest_true], axis=0)

In [53]:
df_test_under['Arrest'].value_counts()

1    1578903
0    1578903
Name: Arrest, dtype: int64

In [54]:
y_under = df_test_under['Arrest']
x_under = df_test_under.drop(['Arrest'], axis = 1)
X_train_under, X_test_under, y_train_under, y_test_under = train_test_split(x_under, y_under, test_size=0.2, random_state=42, stratify=y_under)

In [55]:
model_2 = LogisticRegression()
model_2.fit(X_train_under,y_train_under)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.



LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [56]:
model_2.score(X_test_under,y_test_under)

0.8073823314258932

In [57]:
y_pred_under = model_2.predict(X_test_under)

In [58]:
confusion_matrix(y_test_under,y_pred_under)

array([[298103,  17678],
       [103972, 211809]])

### cross validation undersampling

In [59]:
cv_under = cross_val_score(LogisticRegression(),X_test_under,y_test_under,cv=10)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logis

In [60]:
np.mean(cv_under)

0.807379167003434

### Performance evaluation

In [61]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.87      0.97      0.91    790749
           1       0.88      0.63      0.74    315781

   micro avg       0.87      0.87      0.87   1106530
   macro avg       0.87      0.80      0.83   1106530
weighted avg       0.87      0.87      0.86   1106530



In [62]:
print(classification_report(y_test_under,y_pred_under))

              precision    recall  f1-score   support

           0       0.74      0.94      0.83    315781
           1       0.92      0.67      0.78    315781

   micro avg       0.81      0.81      0.81    631562
   macro avg       0.83      0.81      0.80    631562
weighted avg       0.83      0.81      0.80    631562



## 3.2 Random Forest Classifier 

### Without undersampling

In [63]:
from sklearn.ensemble import RandomForestClassifier

In [64]:
# find best parameters
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV

In [65]:
rf = RandomForestClassifier()

In [66]:
cv_rf = cross_val_score(RandomForestClassifier(),X_test,y_test,cv=10)

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will chan

In [67]:
np.mean(cv_rf)

0.8801288730986803

In [68]:
rf.fit(X_train,y_train)

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.



RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [69]:
rf.score(X_test,y_test)

0.8808455261041274

In [70]:
y_pred_rf = rf.predict(X_test)

In [71]:
confusion_matrix(y_test,y_pred_rf)

array([[772161,  18588],
       [113260, 202521]])

In [72]:
print(classification_report(y_test,y_pred_rf))

              precision    recall  f1-score   support

           0       0.87      0.98      0.92    790749
           1       0.92      0.64      0.75    315781

   micro avg       0.88      0.88      0.88   1106530
   macro avg       0.89      0.81      0.84   1106530
weighted avg       0.88      0.88      0.87   1106530



### With undersampling

In [73]:
rf_under = RandomForestClassifier()

In [74]:
rf_under.fit(X_train_under,y_train_under)

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.



RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [75]:
rf_under.score(X_test_under,y_test_under)

0.8168715027186563

In [76]:
y_pred_rf_under = rf_under.predict(X_test_under)

In [77]:
confusion_matrix(y_test_under,y_pred_rf_under)

array([[290128,  25653],
       [ 90004, 225777]])

In [78]:
print(classification_report(y_test_under,y_pred_rf_under))

              precision    recall  f1-score   support

           0       0.76      0.92      0.83    315781
           1       0.90      0.71      0.80    315781

   micro avg       0.82      0.82      0.82    631562
   macro avg       0.83      0.82      0.81    631562
weighted avg       0.83      0.82      0.81    631562



### Cross validation

In [79]:
cv_rf_under = cross_val_score(RandomForestClassifier(),X_test_under,y_test_under,cv=10)

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will chan

In [80]:
np.mean(cv_rf_under)

0.8146341962775429

## 3.3 Decision tree classifier

### No undersampling

In [81]:
from sklearn.tree import DecisionTreeClassifier

In [82]:
dt = DecisionTreeClassifier()

In [83]:
dt.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [84]:
dt.score(X_test,y_test)

0.8808337776653141

In [85]:
y_pred_dt = dt.predict(X_test)

In [86]:
confusion_matrix(y_test,y_pred_dt)

array([[772377,  18372],
       [113489, 202292]])

In [87]:
print(classification_report(y_test,y_pred_dt))

              precision    recall  f1-score   support

           0       0.87      0.98      0.92    790749
           1       0.92      0.64      0.75    315781

   micro avg       0.88      0.88      0.88   1106530
   macro avg       0.89      0.81      0.84   1106530
weighted avg       0.88      0.88      0.87   1106530



### With undersampling

In [88]:
dt_under = DecisionTreeClassifier()

In [89]:
dt_under.fit(X_train_under,y_train_under)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [90]:
dt_under.score(X_test_under,y_test_under)

0.8169760055228149

In [91]:
y_pred_dt_under = dt_under.predict(X_test_under)

In [92]:
confusion_matrix(y_test_under,y_pred_dt_under)

array([[290821,  24960],
       [ 90631, 225150]])

In [93]:
print(classification_report(y_test_under,y_pred_dt_under))

              precision    recall  f1-score   support

           0       0.76      0.92      0.83    315781
           1       0.90      0.71      0.80    315781

   micro avg       0.82      0.82      0.82    631562
   macro avg       0.83      0.82      0.81    631562
weighted avg       0.83      0.82      0.81    631562



# 4. Test the random forest algorithm with new data

**We will test the accuracy of our best model (random forest) with the new data from chicago police department and see if our predictions match with the reality (arrest or no arrest)** 

In [94]:
# load the recent data from Chicago Police Department:
chicago_crime_new = pd.read_csv('/Users/adrientalbot/Documents/final-project/data/chicago_crime_new.csv')

In [95]:
chicago_crime_new.isnull().sum()

ID                                 0
Case Number                        4
Date                               0
Block                              0
IUCR                               0
Primary Type                       0
Description                        0
Location Description            4823
Arrest                             0
Domestic                           0
Beat                               0
District                          47
Ward                          614822
Community Area                613496
FBI Code                           0
X Coordinate                   63888
Y Coordinate                   63888
Year                               0
Updated On                         0
Latitude                       63888
Longitude                      63888
Location                       63888
Historical Wards 2003-2015     83907
Zip Codes                      64203
Community Areas                81199
Census Tracts                  79052
Wards                          81088
B

In [96]:
chicago_crime_new = chicago_crime_new.drop(['ID', 'Case Number', 'Block', 'Ward', 'Community Area', 'FBI Code',
                                 'X Coordinate', 'Y Coordinate', 'Updated On'], axis = 1)

In [97]:
chicago_crime_new = chicago_crime_new.dropna()

In [98]:
chicago_crime_new = chicago_crime_new.set_index(chicago_crime_new.Date)

In [99]:
chicago_crime_new.head()

Date  IUCR         Primary Type  \
Date                                                                        
04/15/2019 07:55:00 PM  04/15/2019 07:55:00 PM  0910  MOTOR VEHICLE THEFT   
04/15/2019 07:30:00 PM  04/15/2019 07:30:00 PM  1320      CRIMINAL DAMAGE   
04/15/2019 07:30:00 PM  04/15/2019 07:30:00 PM  143B    WEAPONS VIOLATION   
04/15/2019 07:27:00 PM  04/15/2019 07:27:00 PM  0860                THEFT   
04/15/2019 07:07:00 PM  04/15/2019 07:07:00 PM  2024            NARCOTICS   

                                        Description Location Description  \
Date                                                                       
04/15/2019 07:55:00 PM                   AUTOMOBILE               STREET   
04/15/2019 07:30:00 PM                   TO VEHICLE               STREET   
04/15/2019 07:30:00 PM  UNLAWFUL POSS OTHER FIREARM               STREET   
04/15/2019 07:27:00 PM                 RETAIL THEFT     DEPARTMENT STORE   
04/15/2019 07:07:00 PM          POSS: HEROIN(WHITE)             SIDEWALK   

                        Arrest  Domestic  Beat  District  Year   Latitude  \
Date                                                                        
04/15/2019 07:55:00 PM   False     False  1834      18.0  2019  41.894951   
04/15/2019 07:30:00 PM   False     False  1631      16.0  2019  41.947005   
04/15/2019 07:30:00 PM    True     False   724       7.0  2019  41.771950   
04/15/2019 07:27:00 PM    True     False  1733      17.0  2019  41.946649   
04/15/2019 07:07:00 PM    True     False   933       9.0  2019  41.802450   

                        Longitude                       Location  \
Date                                                               
04/15/2019 07:55:00 PM -87.623644  (41.894950514, -87.623644357)   
04/15/2019 07:30:00 PM -87.807066  (41.947005032, -87.807066054)   
04/15/2019 07:30:00 PM -87.649442  (41.771949874, -87.649442482)   
04/15/2019 07:27:00 PM -87.702085  (41.946648722, -87.702084879)   
04/15/2019 07:07:00 PM -87.658731   (41.80244993, -87.658731165)   

                        Historical Wards 2003-2015  Zip Codes  \
Date                                                            
04/15/2019 07:55:00 PM                        22.0    21182.0   
04/15/2019 07:30:00 PM                        39.0    22254.0   
04/15/2019 07:30:00 PM                        17.0    21559.0   
04/15/2019 07:27:00 PM                         2.0    21538.0   
04/15/2019 07:07:00 PM                        19.0    14924.0   

                        Community Areas  Census Tracts  Wards  \
Date                                                            
04/15/2019 07:55:00 PM             37.0          158.0   36.0   
04/15/2019 07:30:00 PM             17.0          392.0   19.0   
04/15/2019 07:30:00 PM             66.0          410.0   32.0   
04/15/2019 07:27:00 PM             22.0          425.0   20.0   
04/15/2019 07:07:00 PM             59.0          739.0    4.0   

                        Boundaries - ZIP Codes  Police Districts  Police Beats  
Date                                                                            
04/15/2019 07:55:00 PM                     6.0              14.0          74.0  
04/15/2019 07:30:00 PM                    52.0              12.0          29.0  
04/15/2019 07:30:00 PM                    11.0              17.0         202.0  
04/15/2019 07:27:00 PM                    39.0               1.0          23.0  
04/15/2019 07:07:00 PM                    37.0              23.0         118.0

**We will use the training data from the original dataset (df_ml)**

In [100]:
# use df_test_under if you want undersampled data
X_train_original = ml_df.drop('Arrest', axis = 1)
y_train_original = ml_df['Arrest']

**We will use the new downloaded dataset (chicago_crime_new) as our test data. 
But first we have to clean the new dataset to make it useable for the machine learning algorithm**

In [101]:
chicago_crime_new_formatted = chicago_crime_new.drop(columns = ['Latitude','Longitude', 'Date','IUCR','Description','Beat','Location','Historical Wards 2003-2015','Zip Codes', 'Community Areas','Census Tracts','Wards','Boundaries - ZIP Codes','Police Districts','Police Beats','Year'], axis = 1)

In [102]:
def conversion(x):
    if x == True:
        return 1
    else: return 0  

In [103]:
chicago_crime_new_formatted['Arrest'] = chicago_crime_new_formatted['Arrest'].apply(lambda x: conversion(x))
chicago_crime_new_formatted['Domestic'] = chicago_crime_new_formatted['Domestic'].apply(lambda x: conversion(x))

In [104]:
chicago_crime_new_formatted.head()

Primary Type Location Description  Arrest  \
Date                                                                       
04/15/2019 07:55:00 PM  MOTOR VEHICLE THEFT               STREET       0   
04/15/2019 07:30:00 PM      CRIMINAL DAMAGE               STREET       0   
04/15/2019 07:30:00 PM    WEAPONS VIOLATION               STREET       1   
04/15/2019 07:27:00 PM                THEFT     DEPARTMENT STORE       1   
04/15/2019 07:07:00 PM            NARCOTICS             SIDEWALK       1   

                        Domestic  District  
Date                                        
04/15/2019 07:55:00 PM         0      18.0  
04/15/2019 07:30:00 PM         0      16.0  
04/15/2019 07:30:00 PM         0       7.0  
04/15/2019 07:27:00 PM         0      17.0  
04/15/2019 07:07:00 PM         0       9.0

In [105]:
#one hot encoding for primary types: 
chicago_formatted_new_primary = chicago_crime_new_formatted[chicago_crime_new_formatted['Primary Type'].isin(lst_primary_types)]
for primary_type in lst_primary_types:
    chicago_formatted_new_primary[primary_type] = 0
for primary_type in lst_primary_types:
    chicago_formatted_new_primary[primary_type] = np.where(chicago_formatted_new_primary['Primary Type'].str.contains(primary_type), 1, 0)

In [106]:
#one hot encoding for location types: 
chicago_formatted_new_primary_location = chicago_formatted_new_primary[chicago_formatted_new_primary['Location Description'].isin(lst_location)]
for location in lst_location:
    chicago_formatted_new_primary_location[location] = 0
for location in lst_location:
    chicago_formatted_new_primary_location[location] = np.where(chicago_formatted_new_primary_location['Location Description'].str.contains(location), 1, 0)


/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: UserWarning:

This pattern has match groups. To actually get the groups, use str.extract.



In [107]:
one_hot_table_test = chicago_formatted_new_primary_location.drop(['Primary Type','Location Description'], axis =1)
one_hot_table_test.head()

Arrest  Domestic  District  THEFT  \
Date                                                        
04/15/2019 07:55:00 PM       0         0      18.0      1   
04/15/2019 07:30:00 PM       0         0      16.0      0   
04/15/2019 07:30:00 PM       1         0       7.0      0   
04/15/2019 07:27:00 PM       1         0      17.0      1   
04/15/2019 07:07:00 PM       1         0       9.0      0   

                        OFFENSE INVOLVING CHILDREN  ASSAULT  \
Date                                                          
04/15/2019 07:55:00 PM                           0        0   
04/15/2019 07:30:00 PM                           0        0   
04/15/2019 07:30:00 PM                           0        0   
04/15/2019 07:27:00 PM                           0        0   
04/15/2019 07:07:00 PM                           0        0   

                        CRIM SEXUAL ASSAULT  SEX OFFENSE  DECEPTIVE PRACTICE  \
Date                                                                           
04/15/2019 07:55:00 PM                    0            0                   0   
04/15/2019 07:30:00 PM                    0            0                   0   
04/15/2019 07:30:00 PM                    0            0                   0   
04/15/2019 07:27:00 PM                    0            0                   0   
04/15/2019 07:07:00 PM                    0            0                   0   

                        CRIMINAL DAMAGE  BATTERY  MOTOR VEHICLE THEFT  \
Date                                                                    
04/15/2019 07:55:00 PM                0        0                    1   
04/15/2019 07:30:00 PM                1        0                    0   
04/15/2019 07:30:00 PM                0        0                    0   
04/15/2019 07:27:00 PM                0        0                    0   
04/15/2019 07:07:00 PM                0        0                    0   

                        BURGLARY  WEAPONS VIOLATION  CRIMINAL TRESPASS  \
Date                                                                     
04/15/2019 07:55:00 PM         0                  0                  0   
04/15/2019 07:30:00 PM         0                  0                  0   
04/15/2019 07:30:00 PM         0                  1                  0   
04/15/2019 07:27:00 PM         0                  0                  0   
04/15/2019 07:07:00 PM         0                  0                  0   

                        ROBBERY  NARCOTICS  LIQUOR LAW VIOLATION  \
Date                                                               
04/15/2019 07:55:00 PM        0          0                     0   
04/15/2019 07:30:00 PM        0          0                     0   
04/15/2019 07:30:00 PM        0          0                     0   
04/15/2019 07:27:00 PM        0          0                     0   
04/15/2019 07:07:00 PM        0          1                     0   

                        PUBLIC PEACE VIOLATION  \
Date                                             
04/15/2019 07:55:00 PM                       0   
04/15/2019 07:30:00 PM                       0   
04/15/2019 07:30:00 PM                       0   
04/15/2019 07:27:00 PM                       0   
04/15/2019 07:07:00 PM                       0   

                        INTERFERENCE WITH PUBLIC OFFICER  HOMICIDE  GAMBLING  \
Date                                                                           
04/15/2019 07:55:00 PM                                 0         0         0   
04/15/2019 07:30:00 PM                                 0         0         0   
04/15/2019 07:30:00 PM                                 0         0         0   
04/15/2019 07:27:00 PM                                 0         0         0   
04/15/2019 07:07:00 PM                                 0         0         0   

                        KIDNAPPING  PROSTITUTION  ARSON  STREET  RESIDENCE  \
Date                                                                         
04/15/2019 07:55:

In [108]:
# arrest rate/district column
chicago_crime_counts_test = pd.DataFrame(one_hot_table_test['District'].value_counts()).reset_index()
chicago_crime_counts_test = chicago_crime_counts_test.rename(columns={'index': 'District', 'District': 'Crime_Counts'})
chicago_arrest_counts_test = pd.DataFrame(one_hot_table_test[one_hot_table_test['Arrest'] == 1]['District'].value_counts()).reset_index()
chicago_arrest_counts_test = chicago_arrest_counts_test.rename(columns={'index': 'District', 'District': 'Crime_Counts'})
chicago_map_test = chicago_crime_counts_test.merge(chicago_arrest_counts_test, on='District', how='left')
chicago_map_test = chicago_map_test.rename({'Crime_Counts_x': 'total crimes', 'Crime_Counts_y': 'total arrests'}, axis = 1)
chicago_map_test['arrest rate'] = chicago_map_test['total arrests'] / chicago_map_test['total crimes']
ml_df_test_2 = one_hot_table_test.merge(chicago_map_test, on = 'District', how ='left')
ml_df_test_2 = ml_df_test_2.drop(['total crimes' ,'total arrests', 'District'], axis = 1)


In [109]:
ml_df_test_2.to_csv(r'/Users/adrientalbot/Documents/final-project/data/dataset_ml_test.csv')

In [110]:
X_test_pred = ml_df_test_2.drop('Arrest', axis = 1)
y_test_pred = ml_df_test_2['Arrest']

**We fit the algorithm with training data from the original data set and we will test the algorithm on the new dataset.**

In [111]:
random_forest_pred = RandomForestClassifier()
random_forest_pred.fit(X_train_original , y_train_original)

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning:

The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.



RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [112]:
preds = random_forest_pred.predict(X_test_pred)

In [113]:
chicago_formatted_new_primary_location['Predictions'] = preds

In [114]:
predictions = chicago_formatted_new_primary_location.drop(lst_primary_types, axis = 1)
predictions  = predictions.drop(lst_location, axis = 1).reset_index()
predictions_1504 = predictions[predictions['Date'].str.contains('04/15/2019')]

In [124]:
predictions_1504['District'] = predictions_1504['District'].astype('int64')
predictions_1504.head()

Date         Primary Type Location Description  Arrest  \
0  04/15/2019 07:55:00 PM  MOTOR VEHICLE THEFT               STREET       0   
1  04/15/2019 07:30:00 PM      CRIMINAL DAMAGE               STREET       0   
2  04/15/2019 07:30:00 PM    WEAPONS VIOLATION               STREET       1   
3  04/15/2019 07:27:00 PM                THEFT     DEPARTMENT STORE       1   
4  04/15/2019 07:07:00 PM            NARCOTICS             SIDEWALK       1   

   Domestic  District  Predictions  
0         0        18            0  
1         0        16            0  
2         0         7            1  
3         0        17            1  
4         0         9            1

In [116]:
#predictions_1504.to_csv(r'/Users/adrientalbot/Documents/final-project/data/predictions_april_2019.csv')

In [117]:
pd.crosstab(predictions_1504['Arrest'], predictions_1504['Predictions'])

Predictions    0   1
Arrest              
0            143  12
1             21  41

**As you can see, our random forest algorithm predicts correctly whether a crime does not lead an arrest (143 correct out of 155). However we can see that out of 66 crimes that were actually arrests, our algorithm predicted only 41 of those.** 